pkg/urbit/jets/e/cue.c

```diff
@@ -3,94 +3,204 @@
 */
 #include "all.h"
 
+typedef struct cueframe
+{
+  c3_y    tag_y;
+  u3_atom cur;
+  u3_atom wid;
+  u3_noun hed;
+} cueframe;
+
+static inline void
+_cue_push(c3_ys   mov,
+          c3_ys   off,
+          c3_y    tag_y,
+          u3_atom cur,
+          u3_atom wid,
+          u3_noun hed)
+{
+  u3R->cap_p += mov;
+
+  //  ensure we haven't overflowed the stack
+  //  (off==0 means we're on a north road)
+  //
+  if ( 0 == off ) {
+    c3_assert(u3R->cap_p > u3R->hat_p);
+  }
+  else {
+    c3_assert(u3R->cap_p < u3R->hat_p);
+  }
+
+  cueframe* fam_u = u3to(cueframe, u3R->cap_p + off);
+  fam_u->tag_y = tag_y;
+  fam_u->cur   = cur;
+  fam_u->wid   = wid;
+  fam_u->hed   = hed;
+}
+
+static inline cueframe
+_cue_pop(c3_ys mov, c3_ys off)
+{
+  cueframe* fam_u = u3to(cueframe, u3R->cap_p + off);
+  u3R->cap_p -= mov;
+
+  return *fam_u;
+}
+
+#define CUE_ROOT 0
+#define CUE_HEAD 1
+#define CUE_TAIL 2
+
+u3_noun
+u3qe_cue(u3_atom a)
+{
+  c3_ys mov, off;
 
-  static u3_noun
-  _cue_in(u3p(u3h_root) har_p,
-          u3_atom       a,
-          u3_atom       b)
   {
-    u3_noun p, q;
+    c3_y wis_y = c3_wiseof(cueframe);
+    c3_o nor_o = u3a_is_north(u3R);
+    mov = ( c3y == nor_o ? -wis_y : wis_y );
+    off = ( c3y == nor_o ? 0 : -wis_y );
+  }
 
-    if ( 0 == u3qc_cut(0, b, 1, a) ) {
-      u3_noun x = u3qa_inc(b);
-      u3_noun c = u3qe_rub(x, a);
+  u3p(u3h_root) har_p = u3h_new();
+
+  //  stash the current stack post
+  //
+  u3p(cueframe) cap_p = u3R->cap_p;
+  _cue_push(mov, off, CUE_ROOT, 0, 0, 0);
+
+  u3_atom cur = 0;
+  u3_atom wid, pro;
+
+  //  read from atom at at cursor
+  //  push on to the stack and continue (cell-head recursion)
+  //  or set .wid and .pro and goto give
+  //  TRANSFER .cur
+  //
+  pass: {
+    //  read tag bit at cur
+    //
+    c3_y tag_y = u3qc_cut(0, cur, 1, a);
+
+    //  low bit unset, cursor points to an atom
+    //
+    if ( 0 == tag_y ) {
+      u3_noun bur;
+      {
+        u3_noun x = u3qa_inc(cur);
+        bur = u3qe_rub(x, a);
+        u3z(x);
+      }
 
-      p = u3qa_inc(u3k(u3h(c)));
-      q = u3k(u3t(c));
+      pro = u3k(u3t(bur));
+      u3h_put(har_p, cur, u3k(pro));
```

Comment:
Why do we `u3k(pro)` here? Didn't we already increment its refcount in the previous line?